File: data_structures/queue/circularqueue_using_array.py
class queue():
	"""docstring for queue"""
	def __init__(self, size=5):
		self.size = size
		self.front = self.rear = -1
		self.array = size*[None]

	def push(self,value):
		if((self.rear==self.size-1 and self.front==0) or (self.rear+1==self.front)):
			print("\nqueue is full.")
			return 0

		elif(self.rear==-1):
			self.rear=self.front=0
			self.array[self.rear]=value

		elif self.rear==self.size-1 and self.front>0:
			self.rear=0
			self.array[self.rear]=value

		else:
			self.rear+=1
			self.array[self.rear]=value

		print("\n",value, " is added in queue")
		print("\n front: ",self.front)
		print("\n rear: ",self.rear)

	def pop(self):
		if(self.front==-1):
			print("\nqueue is empty")
			return 0
		print(self.array[self.front], " is poped")
		if self.front==self.rear:
			self.front = self.rear = -1

		elif self.front==self.size-1 and self.rear>=0:
			self.front=0

		else:
			self.front+=1

		print("\n front: ",self.front)
		print("\n rear: ",self.rear)




qu=queue(5)
qu.push(1)
qu.push(2)
qu.push(3)
qu.pop()
qu.push(4)
qu.push(5)
qu.push(6)
qu.pop()
qu.pop()
qu.pop()
qu.push(7)
qu.pop()
qu.pop()
qu.pop()
qu.push(4)
qu.push(5)
qu.push(6)
qu.pop()
qu.pop()

File: data_structures/queue/circularqueue_using_array.cpp
#include <iostream>
using namespace std;
struct queue{
	int size,front,rear,*Q;
};

void creatqueue(struct queue *qu){
	cout<<"\nenter the size of queue: ";
	cin>>qu->size;
	qu->front=qu->rear=-1;
	qu->Q=new int[qu->size];
}

int push(struct queue *qu,int value){
	if(qu->rear-qu->front==qu->size-1 || qu->rear+1==qu->front){
		cout<<"\nqueue is full";
		cout<<"\nfront : "<<qu->front;
		cout<<"\nrear: "<<qu->rear;
		return 0;
	}
	else if(qu->rear==-1){
		qu->rear=qu->front=0;
		qu->Q[qu->rear]=value;
	}
	else if(qu->rear==qu->size-1 && qu->front>=0){
		qu->rear=0;
		qu->Q[qu->rear]=value;
	}
	else{
		qu->rear++;
		qu->Q[qu->rear]=value;
	}
	cout<<"\n"<<value<<" is added in queue";
	cout<<"\nfront : "<<qu->front;
	cout<<"\nrear: "<<qu->rear;
	return value;
}

int pop(struct queue *qu){
	int x=0;
	if(qu->front==-1){
		cout<<"\nqueue is empty";
		return x;
	}
	else if(qu->front==qu->rear){
		x=qu->Q[qu->front];
		qu->front=qu->rear=-1;
	}
	else if(qu->front==qu->size-1&&qu->rear>=0){
		x=qu->Q[qu->front];
		qu->front=0;
	}
	else{
		x=qu->Q[qu->front];
		qu->front++;

	}
	cout<<"\n"<<x<<" is poped";
	cout<<"\nfront : "<<qu->front;
	cout<<"\nrear: "<<qu->rear;
	return x;

}

void display(struct queue *qu){
	if(qu->front==-1){
		cout<<"\nqueue is empty";
	}
	else if(qu->rear>=qu->front){
		for(int i=qu->front;i<=qu->rear;i++){
		cout<<"\n"<<qu->Q[i];
	}
	}
	else{
		for(int i=qu->front;i<=qu->size-1;i++){
		cout<<"\n"<<qu->Q[i];}
		for(int i=0;i<=qu->rear;i++){
		cout<<"\n"<<qu->Q[i];
	}

	}

}

int main()
{
	struct queue qu;
	creatqueue(&qu);
	cout<<"\npress 1 for push\npress 2 for pop\npress 3 for display\npress 4 for exit\n";
	while(true){
		cout<<"\nenter choice: ";
		int i;
		cin>>i;
		if (i==1)
		{
			int value;
			cout<<"enter the value: ";
			cin>>value;
			push(&qu,value);
		}
		else if (i==2)
		{
			pop(&qu);
		}
		else if (i==3)
		{
			display(&qu);
		}
		else
		{
			break;
		}
	}
	return 0;
}



